CUDADataFormats/Track/interface/TrackSoAHeterogeneousT.h

```diff
@@ -19,39 +19,61 @@ class TrackSoAHeterogeneousT {
   using hindex_type = uint32_t;
   using HitContainer = cms::cuda::OneToManyAssoc<hindex_type, S, 5 * S>;
 
+
+  // quality accessors
+  constexpr Quality quality(int32_t i) const { return reinterpret_cast<Quality>(quality_(i)); }
+  constexpr Quality &quality(int32_t i) { return reinterpret_cast<Quality &>(quality_(i)); }
+  constexpr Quality const *qualityData() const { return reinterpret_cast<Quality const *>(quality_.data()); }
+  constexpr Quality *qualityData() { return reinterpret_cast<Quality *>(quality_.data()); }
+
+  // chi2 accessors
+  constexpr auto & chi2(int32_t i) { return chi2_(i); }
+  constexpr auto chi2(int32_t i) const { return chi2_(i); }
+
+  // stateAtBS accessors
+  constexpr auto & stateAtBS() { return stateAtBS_; }
+  constexpr auto stateAtBS() const { return stateAtBS_; }
```

Comment:
are we sure to return it by value?

Reply:
You are right, probably not a good idea.

From a quick skim I thought these were scalars... for complex types, better return by `const &` (and please make the return type explicit, not `auto`).

Reply:
Returning a const ref failed with this compilation error:
```
/data/user/cano/PR31722-bis/New/CMSSW_11_3_0_pre5/src/CUDADataFormats/Track/interface/TrackSoAHeterogeneousT.h:44:60: error: returning reference to temporary [-Werror=return-local-addr]
   44 |   constexpr auto const & pt(int32_t i) const { return pt_(i); }
```

It gets fixed by changing cnost eigen SoA operators like this:
```diff
diff --git a/HeterogeneousCore/CUDAUtilities/interface/eigenSoA.h b/HeterogeneousCore/CUDAUtilities/interface/eigenSoA.h
index 021402c93c2..186694bc1ec 100644
--- a/HeterogeneousCore/CUDAUtilities/interface/eigenSoA.h
+++ b/HeterogeneousCore/CUDAUtilities/interface/eigenSoA.h
@@ -19,9 +19,9 @@ namespace eigenSoA {
     using Scalar = T;

     __host__ __device__ constexpr Scalar& operator()(int32_t i) { return data_[i]; }
-    __device__ constexpr const Scalar operator()(int32_t i) const { return __ldg(data_ + i); }
+    __host__ __device__ constexpr Scalar const & operator()(int32_t i) const { return data_[i]; }
     __host__ __device__ constexpr Scalar& operator[](int32_t i) { return data_[i]; }
-    __device__ constexpr const Scalar operator[](int32_t i) const { return __ldg(data_ + i); }
+    __host__ __device__ constexpr Scalar const & operator[](int32_t i) const { return data_[i]; }

     __host__ __device__ constexpr Scalar* data() { return data_; }
     __host__ __device__ constexpr Scalar const* data() const { return data_; }
```
(which is more natural). Is this change fine too?

Reply:
no is NOT!
please keep __ldg

Reply:
only ```stateAtBS_`` shall be returned by const ref. all others are scalar and shall be returned by value (for the very good reason you just discovered)

Reply:
OK. I will just do the change with `stateAtBS` (as was requested). I thought the whole block needed the change.